pilot/pkg/security/authz/builder/extauthz.go

```diff
@@ -0,0 +1,315 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builder
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/davecgh/go-spew/spew"
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	extauthzhttp "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/ext_authz/v3"
+	extauthztcp "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/ext_authz/v3"
+	envoy_type_matcher_v3 "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+	envoytypev3 "github.com/envoyproxy/go-control-plane/envoy/type/v3"
+	"github.com/golang/protobuf/ptypes/duration"
+	"github.com/hashicorp/go-multierror"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	authzmodel "istio.io/istio/pilot/pkg/security/authz/model"
+	"istio.io/istio/pkg/config/host"
+)
+
+const (
+	extAuthzMatchPrefix = "istio-ext-authz"
+)
+
+var (
+	rbacDefaultDenyAll = &rbacpb.RBAC{
+		Action: rbacpb.RBAC_DENY,
+		Policies: map[string]*rbacpb.Policy{
+			"default-deny-all-due-to-bad-CUSTOM-action": {
+				Permissions: []*rbacpb.Permission{{Rule: &rbacpb.Permission_Any{Any: true}}},
+				Principals:  []*rbacpb.Principal{{Identifier: &rbacpb.Principal_Any{Any: true}}},
+			},
+		},
+	}
+)
+
+type builtExtAuthz struct {
+	http *extauthzhttp.ExtAuthz
+	tcp  *extauthztcp.ExtAuthz
+}
+
+func hasDuplicate(names []string) bool {
+	for i := 1; i < len(names); i++ {
+		if names[i-1] != names[i] {
+			return true
+		}
+	}
+	return false
+}
+
+func buildExtAuthz(configs []*meshconfig.MeshConfig_ExtensionProvider, providers []string) (*builtExtAuthz, error) {
+	if hasDuplicate(providers) {
+		return nil, fmt.Errorf("only 1 extension provider is allowed for a workload, found different providers: %v", providers)
```

Comment:
The condition is checking whether names are duplicated, but the error msg is about "only 1 allowed"?

Reply:
see above.